# Exploratory Data Analysis

This is a exploratory data analysis of the cars.csv dataset. The goal is to convert data from the initial format to a format that may be better for analysis. Please note that all code is my own unless otherwise specified.

In [2]:
import pandas as pd

Read data from csv, specifying header as first row.

In [3]:
df = pd.read_csv("cars.csv", header=0)

Use head to view first five rows of the data.

In [4]:
df.head()

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                140                143               202   
1                140                143               202   
2                140                143               202   
3                140                143               202   
4                140                143               202   

  Engine Information.Driveline                Engine Information.Engine Type  \
0              All-wheel drive          Audi 3.2L 6 cylinder 250hp 236ft-lbs   
1            Front-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
2            Front-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
3              All-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
4              All-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   

   Engine Information.Hybrid  Engine Information.Number of Forward Gears  \
0                       True                                           6   
1                       True                                           6   
2                       True                                           6   
3                       True                                           6   
4                       True                                           6   

  Engine Information.Transmission  Fuel Information.City mpg  \
0  6 Speed Automatic Select Shift                         18   
1  6 Speed Automatic Select Shift                         22   
2                  6 Speed Manual                         21   
3  6 Speed Automatic Select Shift                         21   
4  6 Speed Automatic Select Shift                         21   

  Fuel Information.Fuel Type  Fuel Information.Highway mpg  \
0                   Gasoline                            25   
1                   Gasoline                            28   
2                   Gasoline                            30   
3                   Gasoline                            28   
4                   Gasoline                            28   

  Identification.Classification           Identification.ID  \
0        Automatic transmission            2009 Audi A3 3.2   
1        Automatic transmission       2009 Audi A3 2.0 T AT   
2           Manual transmission          2009 Audi A3 2.0 T   
3        Automatic transmission  2009 Audi A3 2.0 T Quattro   
4        Automatic transmission  2009 Audi A3 2.0 T Quattro   

  Identification.Make Identification.Model Year  Identification.Year  \
0                Audi              2009 Audi A3                 2009   
1                Audi              2009 Audi A3                 2009   
2                Audi              2009 Audi A3                 2009   
3                Audi              2009 Audi A3                 2009   
4                Audi              2009 Audi A3                 2009   

   Engine Information.Engine Statistics.Horsepower  \
0                                              250   
1                                              200   
2                                              200   
3                                              200   
4                                              200   

   Engine Information.Engine Statistics.Torque  
0                                          236  
1                                          207  
2                                          207  
3                                          207  
4                                          207

Check the shape of the data

In [5]:
df.shape

(5076, 18)

Check columns

In [6]:
df.columns

Index(['Dimensions.Height', 'Dimensions.Length', 'Dimensions.Width',
       'Engine Information.Driveline', 'Engine Information.Engine Type',
       'Engine Information.Hybrid',
       'Engine Information.Number of Forward Gears',
       'Engine Information.Transmission', 'Fuel Information.City mpg',
       'Fuel Information.Fuel Type', 'Fuel Information.Highway mpg',
       'Identification.Classification', 'Identification.ID',
       'Identification.Make', 'Identification.Model Year',
       'Identification.Year',
       'Engine Information.Engine Statistics.Horsepower',
       'Engine Information.Engine Statistics.Torque'],
      dtype='object')

Use dtypes to check the data types of the columns

In [7]:
df.dtypes

Dimensions.Height                                   int64
Dimensions.Length                                   int64
Dimensions.Width                                    int64
Engine Information.Driveline                       object
Engine Information.Engine Type                     object
Engine Information.Hybrid                            bool
Engine Information.Number of Forward Gears          int64
Engine Information.Transmission                    object
Fuel Information.City mpg                           int64
Fuel Information.Fuel Type                         object
Fuel Information.Highway mpg                        int64
Identification.Classification                      object
Identification.ID                                  object
Identification.Make                                object
Identification.Model Year                          object
Identification.Year                                 int64
Engine Information.Engine Statistics.Horsepower     int64
Engine Informa

From this we can see that some variables are categorical in nature but have object as a dtype so we convert categorical variables such as Identification.Classification, Fuel Information.Fuel Type

In [8]:
df["Identification.Classification"] = pd.Categorical(df["Identification.Classification"])
df["Fuel Information.Fuel Type"] = pd.Categorical(df["Fuel Information.Fuel Type"])
df["Engine Information.Driveline"] = pd.Categorical(df["Engine Information.Driveline"])

Count unique values with nunique

In [9]:
df.nunique()

Dimensions.Height                                   198
Dimensions.Length                                   208
Dimensions.Width                                    138
Engine Information.Driveline                          4
Engine Information.Engine Type                      535
Engine Information.Hybrid                             1
Engine Information.Number of Forward Gears            5
Engine Information.Transmission                      11
Fuel Information.City mpg                            27
Fuel Information.Fuel Type                            4
Fuel Information.Highway mpg                         34
Identification.Classification                         2
Identification.ID                                  5030
Identification.Make                                  47
Identification.Model Year                           918
Identification.Year                                   4
Engine Information.Engine Statistics.Horsepower     216
Engine Information.Engine Statistics.Torque     

We can see that there are 5030 unique IDs but 5076 records so we need to remove duplicate rows based on ID.

In [10]:
df.drop_duplicates(subset='Identification.ID', keep='first', inplace=True)

df.shape

(5030, 18)

Use unique to visualise unique values

In [11]:
for column in df.columns:
    print(column, df[column].unique())

Dimensions.Height [140  91 201 147 172 226 148 112 120  77  22  24  82  85  43  35 167 193
  99 110 118 108 185 195 119 190 203 213 177 175  66  93 247  10 143 152
  45 156 155  58 206 205 191 186 208 139 218 124 109 117 210  13 101 103
 102 104  92  61  19  37 188 179 194 224 132  88  38   1  23 189 168 223
 137 198 180 164 115 135 141 159 170  33 150  96 165 200 162 228 238 157
 144  74 239 127  36 220 173 163 126 158  41  52  54   5  89 169 212 214
 114 244 142 134  71  94 231 111 221 216 138  64 236 183  46 254 125  26
 252 229   3 233  17 246  28 178  11  57  79 196 181   9  27 230 222  20
  30 128 166 255 161 184 197  60 160  12 207  51  25  55  34 211 242   8
 249  95  69  63  80  48  40 146  73   6  39  16 130 153 107 100  78 217
 145  97   7 133   2 234  49  62  42  44  50 105  75 199 113 174  32 241]
Dimensions.Length [143  17 221  96  63 243  12  22 140  60 102 215  52  75 168 155  34 195
 199 170   7 230 123  19  14 169 216 200 196 240 150  95  57 226  66  71
 198 181  93 2

From this we can see some points for improvement.

- All wheel drive and four wheel drive are the same thing.
- Variables contain more than one piece of information. For example Engine Information.Engine Type contains the make, the number of litres, number of cylinders, horsepower, torque and turbo. It'd be ideal to split these into their own variables.
- There seems to be an error with the hybrid feature as all cars are claimed to be hybrid which is clearly not the case.

Replace four wheel drive with all wheel drive

In [12]:
df.replace("Four-wheel drive","All-wheel drive", inplace=True)

In [13]:
df[df["Engine Information.Driveline"] == "Four-wheel drive"]

Empty DataFrame
Columns: [Dimensions.Height, Dimensions.Length, Dimensions.Width, Engine Information.Driveline, Engine Information.Engine Type, Engine Information.Hybrid, Engine Information.Number of Forward Gears, Engine Information.Transmission, Fuel Information.City mpg, Fuel Information.Fuel Type, Fuel Information.Highway mpg, Identification.Classification, Identification.ID, Identification.Make, Identification.Model Year, Identification.Year, Engine Information.Engine Statistics.Horsepower, Engine Information.Engine Statistics.Torque]
Index: []

Drop hybrid feature.

In [14]:
df.drop("Engine Information.Hybrid", axis = 1, inplace=True)

df

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                   140                143               202   
1                   140                143               202   
2                   140                143               202   
3                   140                143               202   
5                    91                 17                62   
...                 ...                ...               ...   
5071                 13                253               201   
5072                141                249               108   
5073                160                249               108   
5074                200                210               110   
5075                200                 94               110   

     Engine Information.Driveline  \
0                 All-wheel drive   
1               Front-wheel drive   
2               Front-wheel drive   
3                 All-wheel drive   
5                 All-wheel drive   
...                           ...   
5071            Front-wheel drive   
5072              All-wheel drive   
5073              All-wheel drive   
5074             Rear-wheel drive   
5075             Rear-wheel drive   

                      Engine Information.Engine Type  \
0               Audi 3.2L 6 cylinder 250hp 236ft-lbs   
1       Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
2       Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
3       Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
5              Audi 3.2L 6 cylinder 265hp 243 ft-lbs   
...                                              ...   
5071         Honda 3.5L 6 Cylinder 250 hp 253 ft-lbs   
5072  Lamborghini 5.2L 10 cylinder 552 hp 398 ft-lbs   
5073  Lamborghini 5.2L 10 cylinder 552 hp 398 ft-lbs   
5074      BMW 3.0L 6 cylinder 315hp 330 ft-lbs Turbo   
5075      BMW 3.0L 6 cylinder 315hp 330 ft-lbs Turbo   

      Engine Information.Number of Forward Gears  \
0                                              6   
1                                              6   
2                                              6   
3                                              6   
5                                              6   
...                                          ...   
5071                                           5   
5072                                           6   
5073                                           6   
5074                                           6   
5075                                           6   

     Engine Information.Transmission  Fuel Information.City mpg  \
0     6 Speed Automatic Select Shift                         18   
1     6 Speed Automatic Select Shift                         22   
2                     6 Speed Manual                         21   
3     6 Speed Automatic Select Shift                         21   
5                     6 Speed Manual                         16   
...                              ...                        ...   
5071               5 Speed Automatic                         18   
5072                  6 Speed Manual                         12   
5073                  6 Speed Manual                         12   
5074  6 Speed Automatic Select Shift                         17   
5075  6 Speed Automatic Select Shift                         17   

     Fuel Information.Fuel Type  Fuel Information.Highway mpg  \
0                      Gasoline                            25   
1                      Gasoline                            28   
2                      Gasoline                            30   
3                      Gasoline                            28   
5                      Gasoline                            27   
...                         ...                           ...   
5071                   Gasoline                            25   
5072                   Gasoline                            20   
5073                   Gasoline                            20   
5074                   Gasoline        

Split Engine Information.Engine Type into its various entities

In [15]:
df_copy = df.copy(deep=True)
split_columns = df_copy["Engine Information.Engine Type"].str.split(" ", expand=True)

capacity = split_columns[1]
df_copy["engine_make"] = split_columns[0].where(capacity.str[-1].str.upper() == "L", split_columns[0] + " " + capacity)
df_copy["engine_capacity"] = capacity.str[:-1].where(capacity.str[-1].str.upper() == "L", split_columns[2].str[:-1])
df_copy["engine_cylinders"] = split_columns[2].where(capacity.str[-1].str.upper() == "L", split_columns[3])
df_copy["engine_modifier"] = split_columns[8].where(capacity.str[-1].str.upper() == "L", split_columns[9])

# Fill empty cylinders with 6 (Cadillac quirk)
df_copy["engine_cylinders"].replace("", "6", inplace=True)

# Fill empty capacities (Volvo quirk)
mask = df_copy["engine_capacity"] == ""
df_copy.loc[mask, "engine_capacity"] = df_copy.loc[mask, "engine_make"].str.split(" ", expand=True)[1]

# Fix Volvo engine make
df_copy.loc[mask, "engine_make"] = df_copy.loc[mask, "engine_make"].str.split(" ", expand=True)[0]

# Replace 2010 Volkswagen Touareg V6 TDI Clean Diesel Sport cylinders
df_copy["engine_cylinders"].replace("225", "6", inplace=True)

# Replace RX-8 cylinders
df_copy["engine_cylinders"].replace("R2", "4", inplace=True)

# Replace Aston Martin 6.0L 510 hp 420 ft-lbs cylinders
df_copy["engine_cylinders"].replace("510", "12", inplace=True)

# Replace 2012 Lexus LFA cylinders
df_copy["engine_cylinders"].replace("552", "10", inplace=True)

# Drop old column.
df_copy.drop("Engine Information.Engine Type", axis=1, inplace=True)
df_copy

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                   140                143               202   
1                   140                143               202   
2                   140                143               202   
3                   140                143               202   
5                    91                 17                62   
...                 ...                ...               ...   
5071                 13                253               201   
5072                141                249               108   
5073                160                249               108   
5074                200                210               110   
5075                200                 94               110   

     Engine Information.Driveline  Engine Information.Number of Forward Gears  \
0                 All-wheel drive                                           6   
1               Front-wheel drive                                           6   
2               Front-wheel drive                                           6   
3                 All-wheel drive                                           6   
5                 All-wheel drive                                           6   
...                           ...                                         ...   
5071            Front-wheel drive                                           5   
5072              All-wheel drive                                           6   
5073              All-wheel drive                                           6   
5074             Rear-wheel drive                                           6   
5075             Rear-wheel drive                                           6   

     Engine Information.Transmission  Fuel Information.City mpg  \
0     6 Speed Automatic Select Shift                         18   
1     6 Speed Automatic Select Shift                         22   
2                     6 Speed Manual                         21   
3     6 Speed Automatic Select Shift                         21   
5                     6 Speed Manual                         16   
...                              ...                        ...   
5071               5 Speed Automatic                         18   
5072                  6 Speed Manual                         12   
5073                  6 Speed Manual                         12   
5074  6 Speed Automatic Select Shift                         17   
5075  6 Speed Automatic Select Shift                         17   

     Fuel Information.Fuel Type  Fuel Information.Highway mpg  \
0                      Gasoline                            25   
1                      Gasoline                            28   
2                      Gasoline                            30   
3                      Gasoline                            28   
5                      Gasoline                            27   
...                         ...                           ...   
5071                   Gasoline                            25   
5072                   Gasoline                            20   
5073                   Gasoline                            20   
5074                   Gasoline                            25   
5075                   Gasoline                            25   

     Identification.Classification                          Identification.ID  \
0           Automatic transmission                           2009 Audi A3 3.2   
1           Automatic transmission                      2009 Audi A3 2.0 T AT   
2              Manual transmission                         2009 Audi A3 2.0 T   
3           Automatic transmission                 2009 Audi A3 2.0 T Quattro   
5              Manual transmission                           2009 Audi A5 3.2   
...                            ...                                        ...   
5071        Automatic transmission                      2012 Honda Pilot EX-L   
5072           Manual transmis

Remove year from Identification.Model Year since year is already in Identification.Year

In [16]:
model_year = df_copy["Identification.Model Year"].str.split(" ")
model_year = model_year.apply(lambda x: x[1:])

df_copy["Identification.Model Year"] = model_year.apply(lambda x: " ".join(x))
df_copy

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                   140                143               202   
1                   140                143               202   
2                   140                143               202   
3                   140                143               202   
5                    91                 17                62   
...                 ...                ...               ...   
5071                 13                253               201   
5072                141                249               108   
5073                160                249               108   
5074                200                210               110   
5075                200                 94               110   

     Engine Information.Driveline  Engine Information.Number of Forward Gears  \
0                 All-wheel drive                                           6   
1               Front-wheel drive                                           6   
2               Front-wheel drive                                           6   
3                 All-wheel drive                                           6   
5                 All-wheel drive                                           6   
...                           ...                                         ...   
5071            Front-wheel drive                                           5   
5072              All-wheel drive                                           6   
5073              All-wheel drive                                           6   
5074             Rear-wheel drive                                           6   
5075             Rear-wheel drive                                           6   

     Engine Information.Transmission  Fuel Information.City mpg  \
0     6 Speed Automatic Select Shift                         18   
1     6 Speed Automatic Select Shift                         22   
2                     6 Speed Manual                         21   
3     6 Speed Automatic Select Shift                         21   
5                     6 Speed Manual                         16   
...                              ...                        ...   
5071               5 Speed Automatic                         18   
5072                  6 Speed Manual                         12   
5073                  6 Speed Manual                         12   
5074  6 Speed Automatic Select Shift                         17   
5075  6 Speed Automatic Select Shift                         17   

     Fuel Information.Fuel Type  Fuel Information.Highway mpg  \
0                      Gasoline                            25   
1                      Gasoline                            28   
2                      Gasoline                            30   
3                      Gasoline                            28   
5                      Gasoline                            27   
...                         ...                           ...   
5071                   Gasoline                            25   
5072                   Gasoline                            20   
5073                   Gasoline                            20   
5074                   Gasoline                            25   
5075                   Gasoline                            25   

     Identification.Classification                          Identification.ID  \
0           Automatic transmission                           2009 Audi A3 3.2   
1           Automatic transmission                      2009 Audi A3 2.0 T AT   
2              Manual transmission                         2009 Audi A3 2.0 T   
3           Automatic transmission                 2009 Audi A3 2.0 T Quattro   
5              Manual transmission                           2009 Audi A5 3.2   
...                            ...                                        ...   
5071        Automatic transmission                      2012 Honda Pilot EX-L   
5072           Manual transmis

Drop Engine Information.Transmission since useful information is already in Identification.Classification.

In [17]:
df_copy.drop("Engine Information.Transmission", axis=1, inplace=True)
df_copy

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                   140                143               202   
1                   140                143               202   
2                   140                143               202   
3                   140                143               202   
5                    91                 17                62   
...                 ...                ...               ...   
5071                 13                253               201   
5072                141                249               108   
5073                160                249               108   
5074                200                210               110   
5075                200                 94               110   

     Engine Information.Driveline  Engine Information.Number of Forward Gears  \
0                 All-wheel drive                                           6   
1               Front-wheel drive                                           6   
2               Front-wheel drive                                           6   
3                 All-wheel drive                                           6   
5                 All-wheel drive                                           6   
...                           ...                                         ...   
5071            Front-wheel drive                                           5   
5072              All-wheel drive                                           6   
5073              All-wheel drive                                           6   
5074             Rear-wheel drive                                           6   
5075             Rear-wheel drive                                           6   

      Fuel Information.City mpg Fuel Information.Fuel Type  \
0                            18                   Gasoline   
1                            22                   Gasoline   
2                            21                   Gasoline   
3                            21                   Gasoline   
5                            16                   Gasoline   
...                         ...                        ...   
5071                         18                   Gasoline   
5072                         12                   Gasoline   
5073                         12                   Gasoline   
5074                         17                   Gasoline   
5075                         17                   Gasoline   

      Fuel Information.Highway mpg Identification.Classification  \
0                               25        Automatic transmission   
1                               28        Automatic transmission   
2                               30           Manual transmission   
3                               28        Automatic transmission   
5                               27           Manual transmission   
...                            ...                           ...   
5071                            25        Automatic transmission   
5072                            20           Manual transmission   
5073                            20           Manual transmission   
5074                            25        Automatic transmission   
5075                            25        Automatic transmission   

                              Identification.ID Identification.Make  \
0                              2009 Audi A3 3.2                Audi   
1                         2009 Audi A3 2.0 T AT                Audi   
2                            2009 Audi A3 2.0 T                Audi   
3                    2009 Audi A3 2.0 T Quattro                Audi   
5                              2009 Audi A5 3.2                Audi   
...                                         ...                 ...   
5071                      2012 Honda Pilot EX-L               Honda   
5072   2012 Lamborghini Gallardo Coupe LP 560-4         Lamborghini   
5073  2012 Lamborghini Gallardo LP 560-4 Spyder         Lamborg

Rename columns for uniform naming standard.

In [18]:
df_copy.rename(columns={"Dimensions.Height": "height"}, inplace=True)
df_copy.rename(columns={"Dimensions.Length": "length"}, inplace=True)
df_copy.rename(columns={"Dimensions.Width": "width"}, inplace=True)
df_copy.rename(columns={"Engine Information.Driveline": "driveline"}, inplace=True)
df_copy.rename(columns={"Engine Information.Number of Forward Gears": "num_gears"}, inplace=True)
df_copy.rename(columns={"Fuel Information.City mpg": "city_mpg"}, inplace=True)
df_copy.rename(columns={"Fuel Information.Highway mpg": "highway_mpg"}, inplace=True)
df_copy.rename(columns={"Fuel Information.Fuel Type": "fuel_type"}, inplace=True)
df_copy.rename(columns={"Identification.Classification": "transmission"}, inplace=True)
df_copy.rename(columns={"Identification.ID": "id"}, inplace=True)
df_copy.rename(columns={"Identification.Make": "make"}, inplace=True)
df_copy.rename(columns={"Identification.Model Year": "model"}, inplace=True)
df_copy.rename(columns={"Identification.Year": "year"}, inplace=True)
df_copy.rename(columns={"Engine Information.Engine Statistics.Horsepower": "horsepower"}, inplace=True)
df_copy.rename(columns={"Engine Information.Engine Statistics.Torque": "torque"}, inplace=True)

df_copy

height  length  width          driveline  num_gears  city_mpg fuel_type  \
0        140     143    202    All-wheel drive          6        18  Gasoline   
1        140     143    202  Front-wheel drive          6        22  Gasoline   
2        140     143    202  Front-wheel drive          6        21  Gasoline   
3        140     143    202    All-wheel drive          6        21  Gasoline   
5         91      17     62    All-wheel drive          6        16  Gasoline   
...      ...     ...    ...                ...        ...       ...       ...   
5071      13     253    201  Front-wheel drive          5        18  Gasoline   
5072     141     249    108    All-wheel drive          6        12  Gasoline   
5073     160     249    108    All-wheel drive          6        12  Gasoline   
5074     200     210    110   Rear-wheel drive          6        17  Gasoline   
5075     200      94    110   Rear-wheel drive          6        17  Gasoline   

      highway_mpg            transmission  \
0              25  Automatic transmission   
1              28  Automatic transmission   
2              30     Manual transmission   
3              28  Automatic transmission   
5              27     Manual transmission   
...           ...                     ...   
5071           25  Automatic transmission   
5072           20     Manual transmission   
5073           20     Manual transmission   
5074           25  Automatic transmission   
5075           25  Automatic transmission   

                                             id         make  \
0                              2009 Audi A3 3.2         Audi   
1                         2009 Audi A3 2.0 T AT         Audi   
2                            2009 Audi A3 2.0 T         Audi   
3                    2009 Audi A3 2.0 T Quattro         Audi   
5                              2009 Audi A5 3.2         Audi   
...                                         ...          ...   
5071                      2012 Honda Pilot EX-L        Honda   
5072   2012 Lamborghini Gallardo Coupe LP 560-4  Lamborghini   
5073  2012 Lamborghini Gallardo LP 560-4 Spyder  Lamborghini   
5074                        2012 BMW 740i Sedan          BMW   
5075                       2012 BMW 740Li Sedan          BMW   

                            model  year  horsepower  torque  engine_make  \
0                         Audi A3  2009         250     236         Audi   
1                         Audi A3  2009         200     207         Audi   
2                         Audi A3  2009         200     207         Audi   
3                         Audi A3  2009         200     207         Audi   
5                         Audi A5  2009         265     243         Audi   
...                           ...   ...         ...     ...          ...   
5071                  Honda Pilot  2012         250     253        Honda   
5072    Lamborghini Gallardo Coup  2012         552     398  Lamborghini   
5073  Lamborghini Gallardo Spyder  2012         552     398  Lamborghini   
5074                 BMW 7 Series  2012         315     330          BMW   
5075                 BMW 7 Series  2012         315     330          BMW   

     engine_capacity engine_cylinders engine_modifier  
0                3.2                6            None  
1                2.0                4           Turbo  
2                2.0                4           Turbo  
3                2.0                4           Turbo  
5                3.2                6            None  
...              ...              ...             ...  
5071             3.5                6            None  
5072             5.2               10            None  
5073             5.2               10            None  
5074             3.0                6            None  
5075             3.0                6            None  

[5030 rows x 19 columns]

Re-run key commands.

In [19]:
df_copy.dtypes

height                 int64
length                 int64
width                  int64
driveline           category
num_gears              int64
city_mpg               int64
fuel_type           category
highway_mpg            int64
transmission        category
id                    object
make                  object
model                 object
year                   int64
horsepower             int64
torque                 int64
engine_make           object
engine_capacity       object
engine_cylinders      object
engine_modifier       object
dtype: object

In [20]:
for column in df_copy.columns:
    print(column, df_copy[column].unique())

# df_copy.loc[df_copy["engine_cylinders"] == "552"]
# df.loc[df["Identification.ID"] == "2012 Lexus LFA"]

height [140  91 201 147 172 226 148 112 120  77  22  24  82  85  43  35 167 193
  99 110 118 108 185 195 119 190 203 213 177 175  66  93 247  10 143 152
  45 156 155  58 206 205 191 186 208 139 218 124 109 117 210  13 101 103
 102 104  92  61  19  37 188 179 194 224 132  88  38   1  23 189 168 223
 137 198 180 164 115 135 141 159 170  33 150  96 165 200 162 228 238 157
 144  74 239 127  36 220 173 163 126 158  41  52  54   5  89 169 212 214
 114 244 142 134  71  94 231 111 221 216 138  64 236 183  46 254 125  26
 252 229   3 233  17 246  28 178  11  57  79 196 181   9  27 230 222  20
  30 128 166 255 161 184 197  60 160  12 207  51  25  55  34 211 242   8
 249  95  69  63  80  48  40 146  73   6  39  16 130 153 107 100  78 217
 145  97   7 133   2 234  49  62  42  44  50 105  75 199 113 174  32 241]
length [143  17 221  96  63 243  12  22 140  60 102 215  52  75 168 155  34 195
 199 170   7 230 123  19  14 169 216 200 196 240 150  95  57 226  66  71
 198 181  93 242 197 220 202  24 234

Convert relevant dtypes

In [21]:
df_copy["engine_capacity"] = df_copy["engine_capacity"].astype("float64")
df_copy["engine_cylinders"] = df_copy["engine_cylinders"].astype("int64")

df_copy.dtypes

height                 int64
length                 int64
width                  int64
driveline           category
num_gears              int64
city_mpg               int64
fuel_type           category
highway_mpg            int64
transmission        category
id                    object
make                  object
model                 object
year                   int64
horsepower             int64
torque                 int64
engine_make           object
engine_capacity      float64
engine_cylinders       int64
engine_modifier       object
dtype: object

Check for missing values

In [22]:
df_copy.isna().sum() /len(df)*100

height               0.000000
length               0.000000
width                0.000000
driveline            0.000000
num_gears            0.000000
city_mpg             0.000000
fuel_type            0.000000
highway_mpg          0.000000
transmission         0.000000
id                   0.000000
make                 0.000000
model                0.000000
year                 0.000000
horsepower           0.000000
torque               0.000000
engine_make          0.000000
engine_capacity      0.000000
engine_cylinders     0.000000
engine_modifier     68.727634
dtype: float64

Fill missing values with pd.na

In [23]:
df_copy["engine_modifier"].fillna(pd.NA)

df_copy["engine_modifier"]

0        None
1       Turbo
2       Turbo
3       Turbo
5        None
        ...  
5071     None
5072     None
5073     None
5074     None
5075     None
Name: engine_modifier, Length: 5030, dtype: object

Export cleaned data to csv

In [24]:
df_copy.to_csv("cleaned_cars.csv", index=False)

Create copies of df for each MySQL table.

In [25]:
df_car = df_copy[["id", "make", "model", "year", "horsepower", "torque", "engine_make", "engine_capacity", "engine_cylinders", "engine_modifier"]].copy()
df_driveline = df_copy[["id", "driveline"]].copy()
df_transmission = df_copy[["id", "transmission"]].copy()
df_fuel_type = df_copy[["id", "fuel_type"]].copy()
df_mileage = df_copy[["id", "city_mpg", "highway_mpg"]].copy()
df_driveline.rename(columns={"id": "car_id"}, inplace=True)
df_transmission.rename(columns={"id": "car_id"}, inplace=True)
df_fuel_type.rename(columns={"id": "car_id"}, inplace=True)
df_mileage.rename(columns={"id": "car_id"}, inplace=True)

# Save each df to a csv.

df_car.to_csv("car_table.csv", index=False)
df_driveline.index += 1
df_driveline.to_csv("driveline_table.csv", index_label="Index")
df_transmission.index += 1
df_transmission.to_csv("transmission_table.csv", index_label="Index")
df_fuel_type.index += 1
df_fuel_type.to_csv("fuel_type_table.csv", index_label="Index")
df_mileage.index += 1
df_mileage.to_csv("mileage_table.csv", index_label="Index")